### PR TITLE
More profiler improvements

### PR DESCRIPTION
**Pull request by alexrp**

### CODE REVIEW

**Comment by alexrp** [comment]
cc @kumpera @rodrmoya

**Comment by alexrp** [comment]
build

**Comment by alexrp** [comment]
@directhex any idea what's going on with the bots here?

**Comment by borgdylan** [comment]
Builds fail on my machine locally with the master branch code. It has to do with some TLS define.

**Comment by alexrp** [comment]
I'm talking about the Xvfb errors.

**Comment by alexrp** [comment]
build

File: mono/metadata/assembly.c [hunk 1]
```diff
@@ -3254,6 +3254,18 @@ mono_assembly_get_image (MonoAssembly *assembly)
 	return assembly->image;
 }
 
+/**
```
Comment (kumpera) [inline mono/metadata/assembly.c]: This is missing documentation on memory management.

Could the returned pointer be const?

File: mono/metadata/domain.c [hunk 2]
```diff
@@ -1321,6 +1323,12 @@ mono_domain_get_id (MonoDomain *domain)
 	return domain->domain_id;
 }
 
+const char *
+mono_domain_get_friendly_name (MonoDomain *domain)
```
Comment (kumpera) [inline mono/metadata/domain.c]: Missing documentation. Should mention memory management.

File: mono/metadata/domain.c [hunk 3]
```diff
@@ -1496,6 +1504,18 @@ mono_context_get (void)
 	return GET_APPCONTEXT ();
 }
 
+gint32
+mono_context_get_id (MonoAppContext *context)
```
Comment (kumpera) [inline mono/metadata/domain.c]: Missing docs.

The documentation should explain the lifecycle of those ids.

File: mono/metadata/domain.c [hunk 4]
```diff
@@ -1496,6 +1504,18 @@ mono_context_get (void)
...
+{
+	return context->context_id;
+}
+
+gint32
+mono_context_get_domain_id (MonoAppContext *context)
```
Comment (kumpera) [inline mono/metadata/domain.c]: Missing docs.

The documentation should explain the lifecycle of those ids.

File: mono/metadata/profiler.c [hunk 5]
```diff
@@ -381,6 +385,25 @@ mono_profiler_install_appdomain   (MonoProfileAppDomainFunc start_load, MonoProf
 	prof_list->domain_end_unload = end_unload;
 }
 
+void
+mono_profiler_install_appdomain_name (MonoProfileAppDomainFriendlyNameFunc domain_name_cb)
```
Comment (kumpera) [inline mono/metadata/profiler.c]: Once this is merge, can you add this to the release notes in the website?

File: mono/metadata/profiler.h [hunk 6]
```diff
@@ -30,7 +30,8 @@ typedef enum {
 	MONO_PROFILE_MONITOR_EVENTS   = 1 << 17,
 	MONO_PROFILE_IOMAP_EVENTS     = 1 << 18, /* this should likely be removed, too */
 	MONO_PROFILE_GC_MOVES         = 1 << 19,
-	MONO_PROFILE_GC_ROOTS         = 1 << 20
+	MONO_PROFILE_GC_ROOTS         = 1 << 20,
+	MONO_PROFILE_CONTEXT_EVENTS   = 1 << 21
```
Comment (kumpera) [inline mono/metadata/profiler.h]: Add a comma here, it will avoid line damage the next time we add an entry to this enum.

File: mono/metadata/threads.c [hunk 7]
```diff
@@ -2584,6 +2584,24 @@ ves_icall_System_Runtime_Remoting_Contexts_Context_RegisterContext (MonoAppConte
...
+	 * NOTE: Since finalizers are unreliable for the purposes of ensuring
+	 * cleanup in exceptional circumstances, we don't actually do any
+	 * cleanup work here. We instead do this when we iterate the `contexts`
+	 * hash table. The only purpose of this finalizer, at the moment, is to
+	 * notify the profiler.
+	 */
+
+	//g_print ("Releasing context %d in domain %d\n", ctx->context_id, ctx->domain_id);
+
+	mono_profiler_context_unloaded (ctx);
```
Comment (kumpera) [inline mono/metadata/threads.c]: What's the churn rate of contexts? Do we need to worry about the performance ramification of having the icall in the finalizer?

File: mono/profiler/proflog.c [hunk 8]
```diff
@@ -217,20 +217,24 @@ typedef struct _LogBuffer LogBuffer;
  *
  * type metadata format:
  * type: TYPE_METADATA
- * exinfo: flags: TYPE_LOAD_ERR
+ * exinfo: one of: TYPE_END_LOAD, TYPE_END_UNLOAD (optional for TYPE_THREAD and TYPE_DOMAIN)
```
Comment (kumpera) [inline mono/profiler/proflog.c]: What about TYPE_LOAD_ERR?

File: mono/profiler/proflog.c [hunk 9]
```diff
@@ -217,20 +217,24 @@ typedef struct _LogBuffer LogBuffer;
...
- * 	[flags: uleb128] must be 0
  * 	[name: string] image file name
- * if mtype == TYPE_THREAD
- * 	[flags: uleb128] must be 0
+ * if mtype == TYPE_ASSEMBLY
+ * 	[name: string] assembly name
+ * if mtype == TYPE_DOMAIN && exinfo == 0
+ * 	[name: string] domain friendly name
+ * if mtype == TYPE_CONTEXT
+ * 	[domain: sleb128] domain id as pointer
```
Comment (kumpera) [inline mono/profiler/proflog.c]: What does this mean, "domain id as a pointer"?

File: mono/profiler/proflog.c [hunk 10]
```diff
@@ -885,7 +892,7 @@ gc_reference (MonoObject *obj, MonoClass *klass, uintptr_t size, uintptr_t num,
 	int i;
 	uintptr_t last_offset = 0;
 	//const char *name = mono_class_get_name (klass);
-	LogBuffer *logbuffer = ensure_logbuf (20 + num * 8);
+	LogBuffer *logbuffer = ensure_logbuf (1 + 10 + 10 + 10 + 10 + num * (10 + 10));
```
Comment (kumpera) [inline mono/profiler/proflog.c]: All this magic numbers make my head hurt, we should make this into constants
and/or functions.

File: mono/profiler/proflog.c [hunk 11]
```diff
@@ -918,7 +925,7 @@ heap_walk (MonoProfiler *profiler)
 	LogBuffer *logbuffer;
 	if (!do_heap_shot)
 		return;
-	logbuffer = ensure_logbuf (10);
+	logbuffer = ensure_logbuf (1 + 10);
```
Comment (kumpera) [inline mono/profiler/proflog.c]: more magic numbers :(

File: mono/profiler/proflog.c [hunk 12]
```diff
@@ -945,7 +952,7 @@ heap_walk (MonoProfiler *profiler)
 static void
 gc_event (MonoProfiler *profiler, MonoGCEvent ev, int generation) {
 	uint64_t now;
-	LogBuffer *logbuffer = ensure_logbuf (10);
+	LogBuffer *logbuffer = ensure_logbuf (1 + 10 + 10 + 10);
```
Comment (kumpera) [inline mono/profiler/proflog.c]: Repeated magic number, even worse :sob:

File: mono/profiler/proflog.h [hunk 13]
```diff
@@ -5,7 +5,7 @@
 #define LOG_HEADER_ID 0x4D505A01
 #define LOG_VERSION_MAJOR 0
 #define LOG_VERSION_MINOR 4
-#define LOG_DATA_VERSION 10
+#define LOG_DATA_VERSION 12
```
Comment (kumpera) [inline mono/profiler/proflog.h]: We should change this to single version bump.

Given data version 11 will never be in any released mono, it has no value on its own.

File: mono/profiler/proflog.h [hunk 14]
```diff
@@ -16,6 +16,8 @@
  * version 8: added TYPE_RUNTIME and JIT helpers/trampolines
  * version 9: added MONO_PROFILER_CODE_BUFFER_EXCEPTION_HANDLING
  * version 10: added TYPE_COVERAGE
+ * version 11: added thread ID to TYPE_SAMPLE_HIT
+ * version 12: added more load/unload events
```
Comment (kumpera) [inline mono/profiler/proflog.h]: Shouldn't it mention sample hit sorting and that the header tid is now usable for sample hits?

File: mono/profiler/proflog.h [hunk 15]
```diff
@@ -35,16 +37,15 @@ enum {
 	TYPE_HEAP_OBJECT = 2 << 4,
 	TYPE_HEAP_ROOT   = 3 << 4,
 	/* extended type for TYPE_METADATA */
-	TYPE_START_LOAD   = 1 << 4,
```
Comment (kumpera) [inline mono/profiler/proflog.h]: What are those 3 events been removed?

**Comment by kumpera** [comment]
I left a few notes in the commits and it looks really good overall.

It would be nice to get release notes for this puppy after it's merged.

File: mono/metadata/assembly.c (hunk 1 above)
Comment (alexrp) [inline mono/metadata/assembly.c]: > This is missing documentation on memory management.

I didn't include it as the only precedent for mentioning memory management in embedding API documentation seems to be "don't". We can certainly change improve precedent, though.

> Could the returned pointer be const?

Probably.

File: mono/metadata/domain.c (hunk 2 above)
Comment (alexrp) [inline mono/metadata/domain.c]: For this (and the functions you mention below) docs are missing because they didn't seem to be there for other similar embedding APIs in this file. Wasn't sure if that was intentional or not.

File: mono/metadata/domain.c (hunk 3 above)
Comment (alexrp) [inline mono/metadata/domain.c]: > The documentation should explain the lifecycle of those ids.

There's another precedent problem: The docs don't explain anywhere what the lifecycle of domain IDs is in the first place, for example. Happy to document if this is something we *want* to document.

File: mono/metadata/threads.c (hunk 7 above)
Comment (alexrp) [inline mono/metadata/threads.c]: AFAIK for the most part, there's a 1-to-1 correlation between domains and contexts. New contexts are only created if an app explicitly does so, I think. So I wouldn't worry about it.

File: mono/profiler/proflog.c (hunk 8 above)
Comment (alexrp) [inline mono/profiler/proflog.c]: We've never actually emitted this flag so I decided to remove it in order to not mislead consumers of the format into thinking they'll actually get it and therefore relying on it.

Hmm. I should have done this in a separate commit, and we should probably mention it in the format changelog even if it's not technically a format change.

File: mono/profiler/proflog.c (hunk 9 above)
Comment (alexrp) [inline mono/profiler/proflog.c]: It's the domain ID associated with the context (a context always belongs to one domain, while one domain can have many contexts). The "as a pointer" just means we encode it as a pointer value (even though it's a 32-bit integer) to be consistent with the other `TYPE_METADATA` events.

File: mono/profiler/proflog.h (hunk 13 above)
Comment (alexrp) [inline mono/profiler/proflog.h]: OK.

File: mono/profiler/proflog.h (hunk 15 above)
Comment (alexrp) [inline mono/profiler/proflog.h]: The explanation for these is the same as for `TYPE_LOAD_ERR` above.

File: mono/profiler/proflog.h (hunk 14 above)
Comment (alexrp) [inline mono/profiler/proflog.h]: I'm not sure the sorting counts as a format change. Any number of random factors can have an impact on ordering of events in a file; this is just another one, which should be completely transparent to consumers. They should just see better compression.

Regarding thread ID, as I mentioned on Slack, it was a trivial change to add a thread ID field to the `TYPE_SAMPLE_HIT` event, and it made sense to do since we're bumping the format version for all these new metadata events anyway. On the other hand, making the buffer header thread ID usable for these events was a lot of really ugly, hacky work because the profiler wasn't designed to handle the use case of a log buffer having a different thread ID than the actual thread writing to it.

File: mono/profiler/proflog.c (hunk 10 above)
Comment (alexrp) [inline mono/profiler/proflog.c]: How would that help? Those would be constants that are only relevant in the exact event they correspond to. All they would do is move the event size further away from the actual code that writes data to the buffer, which makes it easier again to make the mistake of not updating the event size when adding event fields.

File: mono/metadata/domain.c (hunk 3 above)
Comment (kumpera) [inline mono/metadata/domain.c]: Lifecycle is better defined in terms of what the user can't expect than on what to expect.

For example, we could say IDs are only unique among live domains/contexts and reuse is unspecified.

Anyway, this can be done after this is merged.

File: mono/profiler/proflog.c (hunk 10 above)
Comment (kumpera) [inline mono/profiler/proflog.c]: I still like for us to address this. A bucket of magic numbers do us no good. Specially when they seem to be repeated all over the place.

How about using structs with packing/alignment set 1 and have size functions based on sizeof?
